lib/puppet/pops/types/types.rb

```diff
@@ -815,6 +815,30 @@ def instance?(o, guard = nil)
     end
   end
 
+  def self.new_function(type)
+    @new_function ||= Puppet::Functions.create_loaded_function(:new_enum, type.loader) do
+      local_types do
+        type 'Selector = Variant[Boolean, Integer]'
+        type 'Format   = String[1]'
+      end
+
+      dispatch :from_args do
+        param          'Selector', :selector      # selects one of the Enum's strings
+        optional_param 'Format',   :string_format # how should the string being formatted?
+      end
+
+      def from_args(selector, format = :default)
+        string = case selector
+          when true then @values[1]
+          when false then @values[0]
+          when Integer then @values[selector]
+          end
```

Comment:
We should assert that the selector is between `0 <= selector < @values.length` and `raise ArgumentError` if it's not. Similarly, we should have an `else` clause and raise if since it's not supported. We do something similar when creating floats:

> raise ArgumentError, _("Illegal radix: %{radix}, expected 2, 8, 10, 16, or default") % { radix: radix }